onnx/defs/nn/defs.cc

```diff
@@ -3156,15 +3156,36 @@ ONNX_OPERATOR_SET_SCHEMA(
           auto* num_heads_attr = ctx.getAttribute("num_heads");
           int64_t num_heads = (num_heads_attr != nullptr) ? num_heads_attr->i() : 0;
 
+          // Ensure that num_heads does not control reshaping of input tensor
+          // when input tensor is 4D
+          int64_t is_input_4d = 1;
+          auto* x_tp = ctx.getInputType(0);
+          if ((x_tp == nullptr) || (!x_tp->has_tensor_type()))
+            return false;
+          if (!(x_tp->tensor_type().has_shape())) {
+            return false;
+          }
+          if (x_tp->tensor_type().shape().dim_size() == 4) {
+            num_heads = 0;
+            is_input_4d = 1;
+          } else if (x_tp->tensor_type().shape().dim_size() == 3) {
+            is_input_4d = 0;
+          } else {
+            return false;
+          }
+
           FunctionBuilder builder(functionProto);
           // Set input tensor to the correct shape if input shape is 3D
           // NewShape = [batch_size, sequence_length, num_heads, head_size]
+
+          // Reshape tensor to 4D if input is 3D
           builder.Const1D("Zero1D", (int64_t)0)
               .Const1D("NumHeads", num_heads) // num_heads
               .Const1D("NegOne", (int64_t)(-1)) // head_size, inferred from other dimensions
               .Add("NewShape = Concat <axis = 0> (Zero1D, Zero1D, NumHeads, NegOne)")
-              .Add("XReshaped = Reshape (X, NewShape)");
-          if (num_heads > 0) {
+              .Add("XReshaped = Reshape (X, NewShape)"); // new shape of input tensor: 4D tensor
+
+          if (is_input_4d == 0) {
```

Comment:
Can we move the `NewShape = ...; XReshaped ...` part into the `if` branch? And rename XReshaped to XTransposed to avoid the extra Identity?